### ISSUE TITLE
Network Boundary Shield false postives and heuristics tweaking

### ISSUE THREAD

**Issue by matogolf** [polcak/jsrestrictor#62]
![outlookErr](https://user-images.githubusercontent.com/44405519/85130665-a6393800-b235-11ea-9024-7e207236ef00.JPG)

When accessing Outlook Web via https://outlook.live.com/, there is an error while Network boundary shield is ON:
"Outlook Web can't reach the internet. Check your connection."

After switching Network boundary shield to OFF the issue disappeared.

**Comment by polcak** [comment]
I tried in Firefox, Brave, and Chrome. I was able to access https://outlook.live.com/owa just fine.

Did you receive the notification by JSR? What IP address or domains were mentioned?

**Comment by matogolf** [comment]
Sorry I was not specific enough. 

Issue is on chrome 83.0.4103.106
JSR level is 0. Only Network boundary shield is activated.

Page loads okay but after filling access credentials and logging into account there is issue. At first it looks okay. Page loads all emails and everything as usual but then when I am trying to open any email, pop up notification (Windows 10) pops up and e.g. images in an email are not showing properly.
![outlookErr2](https://user-images.githubusercontent.com/44405519/85138095-82c8ba00-b242-11ea-8bf4-83832225f425.png)

Then when I tried to reload the page:
Original image from first comment and also sometimes this error is shown in standard mode.
![outlookErr4](https://user-images.githubusercontent.com/44405519/85138099-83f9e700-b242-11ea-8200-e3fa14bc55f2.JPG)

This error is shown on reload while using inkognito mode. 
![outlookErr3](https://user-images.githubusercontent.com/44405519/85138097-83615080-b242-11ea-8931-ba7e10678431.JPG)

**Comment by polcak** [comment]
The problem is that Chrome does not have `dns` API that is behind NBS in Firefox. So there are only heuristics based on failed requests. Currently, 10 failed requests block the domain.

Can you have a look to WebDevTools and the Network tab to see some more information about the failed requests? How frequent are they? Are they for the same URL or domain? Are they related to your messages? Do they have something else in common or suspicious?

Even if there is not a quick solution for better heuristics, we can improve is the text of the message.

What about:

"Host outlook.live.com issued to many unsuccessful requests. This may be just an error in your network connectivity or innocent error of the web site. But it can also be a sign of malicious activities such as using your browser as a proxy to scan your local network. It is up to you to decide if you trust the web site and give it an exception from the Network Boundary Scanner using pop up or the option page."

@IanNov: any other idea?

**Comment by IanNov** [comment]
Hmm, yeah. Not much to add. I think it was expected, that this approach may break some websites. 

From what we can see on pictures I suppose there's too many error responses on site's requests, from what I've seen during the testing, these can occur from various reasons and we're not able to track every single one. 

The best bet for now is to probably update the message as you suggested. Reviewing and updating heuristics might help in some cases, but I dont think we'll be able to make it work on every website without losing threat detection capabilities. 

Maybe we could try to distinguish HTTP error codes and build the heuristic based on them. For example I've seen pages like `seznam.cz` to try to login into the email on page load (so without given credentials) and the request resulted in 401 Not Authorized error on every single page load, until I logged in. So, maybe we might try to look at these and work better with concrete errors. I'm not sure how that would affect threat detection capabilities, maybe this approach would create abusable weakness.

**Comment by polcak** [comment]
Message improved in c6614fb

@IanNov: Can you go through the errors at https://github.com/polcak/jsrestrictor/blob/0ff5acc48ea8c58d1638deb89802af5ff16778e6/chrome/http_shield_chrome.js#L41 and propose weights that reflect how much information the error leaks and how common the error is.

So for example 404 adds 1 to the `hostStatistics` at https://github.com/polcak/jsrestrictor/blob/master/chrome/http_shield_chrome.js#L171 while maybe 406 could be a sign of some incompatibility between the client and the server. So I propose something like:

```js
var httpErrorList = {
...
	404:1,
	405:0.5,
	406:0.1,
...
...
			currentHost[targetUrl.hostname]["httpErrors"] += httpErrorList[headers.statusCode];
```

Going back to 401, it leaks that the server exists but once known, there is not much to be learn in respect to the proxy discovery attack. That is the reason we decided to omit 401. However, blocking a server with excessive 401 errors make sense because it might be a distributed password cracking attack. Suppose a malicious server distributes to a connected client passwords and lets the clients do the dirty work of connecting to the victim server. This way, server logs will show the browser proxy IP address and not the real attacker. If we pick weight 0.01 for 401 that means 1000 attempts. Does it make sense to reintroduce 401 and similar errors with a smallish weight? If so, does it make sense to have something like that for Firefox as well?

**Comment by polcak** [comment]
Heuristcs tweaks moved to #76.

**Comment by polcak** [comment]
Merged, available in 0.4.2, discussion on heuristis in #76.
